dependency_cache.go

```diff
@@ -370,19 +381,27 @@ func (DependencyCache) verify(path string, expected string) error {
 	return nil
 }
 
-func (d DependencyCache) overrideDependencySource(urlD *url.URL) {
-	dependencySourceOverride := sherpa.GetEnvWithDefault("BP_DEPENDENCY_SOURCE_OVERRIDE", "")
-	if dependencySourceOverride != "" {
-		d.Logger.Infof("variable BP_DEPENDENCY_SOURCE_OVERRIDE found. overriding download uri.")
-		urlOverride, err := url.ParseRequestURI(dependencySourceOverride)
-		if err == nil {
+func (d DependencyCache) setDependencyMirror(urlD *url.URL) {
+	dependencyMirror := sherpa.GetEnvWithDefault("BP_DEPENDENCY_MIRROR", "")
+	dependencyMirrorPreserveHost := sherpa.GetEnvWithDefault("BP_DEPENDENCY_MIRROR_PRESERVE_HOST", "false")
+	if dependencyMirror != "" {
+		var originalHost string
+
+		d.Logger.Infof("variable BP_DEPENDENCY_MIRROR found. overriding download uri.")
+		urlOverride, err := url.ParseRequestURI(dependencyMirror)
+
+		if strings.EqualFold(dependencyMirrorPreserveHost, "true") && urlD.Hostname() != "" {
+			originalHost = "/" + urlD.Hostname()
+		}
+
+		if strings.EqualFold(urlOverride.Scheme, "https") || strings.EqualFold(urlOverride.Scheme, "file") {
 			urlD.Scheme = urlOverride.Scheme
 			urlD.User = urlOverride.User
 			urlD.Host = urlOverride.Host
-			urlD.Path = urlOverride.Path + urlD.Path
+			urlD.Path = urlOverride.Path + originalHost + urlD.Path
```

Comment:
I'm not sure I understand what's going on here and the `BP_DEPENDENCY_MIRROR_PRESERVE_HOST`. Can you expand on the use case for that and how it works? Is that something specific that some servers require? Thanks

Reply:
The thing is, paths and artifact names are unique only within the scope of one registry/hostname.

Here is a real world example where the same artifact is made available in different registries:
https://repo1.maven.org/maven2/org/apache/maven/apache-maven/4.0.0-alpha-9/apache-maven-4.0.0-alpha-9-bin.tar.gz
https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/4.0.0-alpha-9/apache-maven-4.0.0-alpha-9-bin.tar.gz

If our local mirror acts as a proxy for both "repo1.maven.org" as well as for "repo.maven.apache.org", substituting the hostname only might lead to unknown results because the mirror wouldn't know what upstream registry to target.
By setting BP_DEPENDENCY_MIRROR=https://local-mirror.acme.com and BP_DEPENDENCY_MIRROR_PRESERVE_HOST=true, we could still ensure uniqueness.
Here are the overridden URIs for this case:
https://local-mirror.acme.com/repo1.maven.org/maven2/org/apache/maven/apache-maven/4.0.0-alpha-9/apache-maven-4.0.0-alpha-9-bin.tar.gz
https://local-mirror.acme.com/repo.maven.apache.org/maven2/org/apache/maven/apache-maven/4.0.0-alpha-9/apache-maven-4.0.0-alpha-9-bin.tar.gz

Usually, the combination of path + filename itself is unique across registries and hostname preservation should not be needed. I just thought, one day, two different artifacts might be hosted under the same path/filename by chance.

Reply:
Ah, ok. I see what you're saying. Do you think we should just genericize this and have a `BP_DEPENDENCY_MIRROR_PREFIX` or something like that?

I've been wondering how much we should allow people to change the path of the URL, but this makes sense. I think the name mirror implies that you'll have the same path suffix, but you definitely might need a different prefix.

Reply:
The prefixing option is provided as part of the "BP_DEPENDENCY_MIRROR" at the moment.

Let's assume the original URI is:
https://repo1.maven.org/maven2/org/apache/maven/apache-maven/4.0.0-alpha-9/apache-maven-4.0.0-alpha-9-bin.tar.gz

The current implementation already allows setting a prefix. You can either set scheme://hostname only, or include a prefix using scheme://hostname/prefix. But this prefix part is static.

Scenario A:
With BP_DEPENDENCY_MIRROR=https://local-mirror.acme.com
Would result in: https://local-mirror.acme.com/maven2/org/apache/maven/apache-maven/4.0.0-alpha-9/apache-maven-4.0.0-alpha-9-bin.tar.gz

Scenario B:
With BP_DEPENDENCY_MIRROR=https://local-mirror.acme.com/custom-prefix
Would result in: https://local-mirror.acme.com/custom-prefix/maven2/org/apache/maven/apache-maven/4.0.0-alpha-9/apache-maven-4.0.0-alpha-9-bin.tar.gz

Now, the BP_DEPENDENCY_MIRROR_PRESERVE_HOST=true can be used additionally in both scenarios. The difference to the prefix is, that the hostname part in the path is dynamic. This part changes according to the original download source of each dependency.

If we'd add BP_DEPENDENCY_MIRROR_PRESERVE_HOST=true in both scenarios from above, the results would be:
Scenario A:
https://local-mirror.acme.com/repo1.maven.org/maven2/org/apache/maven/apache-maven/4.0.0-alpha-9/apache-maven-4.0.0-alpha-9-bin.tar.gz
Scenario B:
https://local-mirror.acme.com/custom-prefix/repo1.maven.org/maven2/org/apache/maven/apache-maven/4.0.0-alpha-9/apache-maven-4.0.0-alpha-9-bin.tar.gz

I believe, the current solution leaves room for pretty much any customization.
Would you prefer to have an additional variable "BP_DEPENDENCY_MIRROR_PREFIX", instead of just appending it to the existing "BP_DEPENDENCY_MIRROR" as the path (as in scenario B).

Reply:
OK, I see. I was thinking of `BP_DEPENDENCY_MIRROR` as just the protocol/host/port, but if we're allowing Path to be included and we treat that as a Path Prefix, I'm OK with that. I think we'll just need to document how that works so folks know it's a prefix that's added not the full path. Ultimately, it's less configuration, and I see how it's different. 

I'm just thinking out loud about ways we could reduce the configuration. Less configuration is usually better.

One other idea. What about just having a placeholder in the Path prefix or the URL like `{originalHost}` or something where we do a string replacement if that token is included?

Ex: `BP_DEPENDENCY_MIRROR=https://local-mirror.acme.com/{originalHost}/maven2/org/apache/maven/apache-maven/4.0.0-alpha-9/apache-maven-4.0.0-alpha-9-bin.tar.gz` would result in `https://local-mirror.acme.com/repo1.maven.org/maven2/org/apache/maven/apache-maven/4.0.0-alpha-9/apache-maven-4.0.0-alpha-9-bin.tar.gz`

This could position us well for future requests, as I could see getting requests for other dynamic changes to the URL path.